internal/tuning.go

```diff
@@ -0,0 +1,455 @@
+type CompositeTunerOptions struct {
+	// WorkflowSlotSupplier is the SlotSupplier used for workflow tasks.
+	WorkflowSlotSupplier SlotSupplier
+	// ActivitySlotSupplier is the SlotSupplier used for activity tasks.
+	ActivitySlotSupplier SlotSupplier
+	// LocalActivitySlotSupplier is the SlotSupplier used for local activities.
+	LocalActivitySlotSupplier SlotSupplier
+	// NexusSlotSupplier is the SlotSupplier used for nexus tasks.
+	NexusSlotSupplier SlotSupplier
+}
+
+// NewCompositeTuner creates a WorkerTuner that uses a combination of slot suppliers.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+func NewCompositeTuner(options CompositeTunerOptions) (WorkerTuner, error) {
+	return &CompositeTuner{
+		workflowSlotSupplier:      options.WorkflowSlotSupplier,
+		activitySlotSupplier:      options.ActivitySlotSupplier,
+		localActivitySlotSupplier: options.LocalActivitySlotSupplier,
+		nexusSlotSupplier:         options.NexusSlotSupplier,
+	}, nil
+}
+
+// FixedSizeTunerOptions are the options used by NewFixedSizeTuner.
+type FixedSizeTunerOptions struct {
+	// NumWorkflowSlots is the number of slots available for workflow tasks.
+	NumWorkflowSlots int
+	// NumActivitySlots is the number of slots available for activity tasks.
+	NumActivitySlots int
+	// NumLocalActivitySlots is the number of slots available for local activities.
+	NumLocalActivitySlots int
+	// NumNexusSlots is the number of slots available for nexus tasks.
+	NumNexusSlots int
+}
+
+// NewFixedSizeTuner creates a WorkerTuner that uses fixed size slot suppliers.
+func NewFixedSizeTuner(options FixedSizeTunerOptions) (WorkerTuner, error) {
+	if options.NumWorkflowSlots <= 0 {
+		options.NumWorkflowSlots = defaultMaxConcurrentTaskExecutionSize
+	}
+	if options.NumActivitySlots <= 0 {
+		options.NumActivitySlots = defaultMaxConcurrentActivityExecutionSize
+	}
+	if options.NumLocalActivitySlots <= 0 {
+		options.NumLocalActivitySlots = defaultMaxConcurrentLocalActivityExecutionSize
+	}
+	if options.NumNexusSlots <= 0 {
+		options.NumNexusSlots = defaultMaxConcurrentTaskExecutionSize
+	}
+	wfSS, err := NewFixedSizeSlotSupplier(options.NumWorkflowSlots)
+	if err != nil {
+		return nil, err
+	}
+	actSS, err := NewFixedSizeSlotSupplier(options.NumActivitySlots)
+	if err != nil {
+		return nil, err
+	}
+	laSS, err := NewFixedSizeSlotSupplier(options.NumLocalActivitySlots)
+	if err != nil {
+		return nil, err
+	}
+	nexusSS, err := NewFixedSizeSlotSupplier(options.NumNexusSlots)
+	if err != nil {
+		return nil, err
+	}
+	return &CompositeTuner{
+		workflowSlotSupplier:      wfSS,
+		activitySlotSupplier:      actSS,
+		localActivitySlotSupplier: laSS,
+		nexusSlotSupplier:         nexusSS,
+	}, nil
+}
+
+// FixedSizeSlotSupplier is a slot supplier that will only ever issue at most a fixed number of
+// slots.
+type FixedSizeSlotSupplier struct {
+	// The maximum number of slots that this supplier will ever issue.
+	NumSlots int
```

Comment:
Is there a reason this is exposed when we have `MaxSlots`? A user may think they could modify this number to change the number of slots

Reply:
Good point, hidden.